src/passes/Inlining.cpp

```diff
@@ -298,68 +298,136 @@ struct Updater : public PostWalker<Updater> {
   Module* module;
   std::map<Index, Index> localMapping;
   Name returnName;
+  Type resultType;
   bool isReturn;
   Builder* builder;
   PassOptions& options;
 
+  struct ReturnCallInfo {
+    // The original `return_call` or `return_call_indirect` or `return_call_ref`
+    // with its operands replaced with `local.get`s.
+    Expression* call;
+    // The branch that is serving as the "return" part of the original
+    // `return_call`.
+    Break* branch;
+  };
+
+  // Collect information on return_calls in the inlined body. Each will be
+  // turned into branches out of the original inlined body followed by
+  // non-return version of the original `return_call`, followed by a branch out
+  // to the caller. The branch labels will be filled in at the end of the walk.
+  std::vector<ReturnCallInfo> returnCallInfos;
+
   Updater(PassOptions& options) : options(options) {}
 
   void visitReturn(Return* curr) {
     replaceCurrent(builder->makeBreak(returnName, curr->value));
   }
-  // Return calls in inlined functions should only break out of the scope of
-  // the inlined code, not the entire function they are being inlined into. To
-  // achieve this, make the call a non-return call and add a break. This does
-  // not cause unbounded stack growth because inlining and return calling both
-  // avoid creating a new stack frame.
-  template<typename T> void handleReturnCall(T* curr, Type results) {
-    if (isReturn) {
+
+  template<typename T> void handleReturnCall(T* curr, Signature sig) {
+    if (isReturn || !curr->isReturn) {
       // If the inlined callsite was already a return_call, then we can keep
       // return_calls in the inlined function rather than downgrading them.
       // That is, if A->B and B->C and both those calls are return_calls
       // then after inlining A->B we want to now have A->C be a
       // return_call.
       return;
     }
-    curr->isReturn = false;
-    curr->type = results;
-    // There might still be unreachable children causing this to be unreachable.
-    curr->finalize();
-    if (results.isConcrete()) {
-      replaceCurrent(builder->makeBreak(returnName, curr));
+
+    Break* branch = builder->makeBreak(Name());
+    if (sig.params.size() || curr->template is<CallIndirect>() ||
+        curr->template is<CallRef>()) {
+      // Set the children to locals at the original callsite before the branch.
+      Block* branchBlock = builder->makeBlock();
+      for (Index i = 0; i < sig.params.size(); ++i) {
+        auto var = builder->addVar(getFunction(), sig.params[i]);
+        branchBlock->list.push_back(
+          builder->makeLocalSet(var, curr->operands[i]));
+        curr->operands[i] = builder->makeLocalGet(var, sig.params[i]);
```

Comment:
This can perhaps use [`ir/localize.h`](https://github.com/WebAssembly/binaryen/blob/eae2638a53e778e24e1ed042f2e16aa0b3b127a5/src/ir/localize.h#L47-L71). That would avoid putting things in locals unnecessarily (it will leave things with no effects in place). But if it's more work maybe not worth it.

Reply:
I will look into it.

Reply:
It looks like this would be slightly tricky to use. The result of `ChildLocalizer::getReplacement` may or may not contain the parent (in this case the call), but we need to know where the parent is so we can replace it with a branch. We could add a new `getChildrenReplacment` method that returns a block with only the sets, perhaps.

Reply:
Just pushed this change, PTAL.